esp-hal/src/dma/buffers.rs

```diff
@@ -1580,3 +1582,368 @@ impl DerefMut for DmaLoopBuf {
+/// point to valid while the buffer is being transferred.
+#[cfg_attr(not(aes_dma), expect(unused))]
+pub(crate) unsafe fn prepare_for_rx(
+    descriptors: &mut [DmaDescriptor],
+    #[cfg(psram_dma)] align_buffers: &mut [Option<ManualWritebackBuffer>; 2],
+    mut data: NonNull<[u8]>,
+) -> (NoBuffer, usize) {
+    let chunk_size =
+        BurstConfig::DEFAULT.max_chunk_size_for(unsafe { data.as_ref() }, TransferDirection::In);
+
+    // The data we have to process may not be appropriate for the DMA:
+    // - it may be improperly aligned for PSRAM
+    // - it may not have a length that is a multiple of the external memory block size
+
+    cfg_if::cfg_if! {
+        if #[cfg(psram_dma)] {
+            let data_addr = data.addr().get();
+            let data_in_psram = crate::psram::psram_range().contains(&data_addr);
+        } else {
+            let data_in_psram = false;
+        }
+    }
+
+    let mut descriptors = unwrap!(DescriptorSet::new(descriptors));
+    let data_len = if data_in_psram {
+        cfg_if::cfg_if! {
+            if #[cfg(psram_dma)] {
+                // This could use a better API, but right now we'll have to build the descriptor list by
+                // hand.
+                let consumed_bytes = build_descriptor_list_for_psram(
+                    &mut descriptors,
+                    align_buffers,
+                    data,
+                );
+
+                // Invalidate data written by the DMA
+                unsafe {
+                    crate::soc::cache_invalidate_addr(data_addr as u32, consumed_bytes as u32);
+                }
+
+                consumed_bytes
+            } else {
+                unreachable!()
+            }
+        }
+    } else {
+        // Just set up descriptors as usual
+        let data_len = data.len();
+        unwrap!(descriptors.link_with_buffer(unsafe { data.as_mut() }, chunk_size));
+        unwrap!(descriptors.set_tx_length(data_len, chunk_size));
+
+        data_len
+    };
+
+    for desc in descriptors.linked_iter_mut() {
+        desc.reset_for_rx();
+    }
+
+    (
+        NoBuffer(Preparation {
+            start: descriptors.head(),
+            direction: TransferDirection::In,
+            burst_transfer: BurstConfig::DEFAULT,
+            check_owner: None,
+            auto_write_back: true,
+            #[cfg(psram_dma)]
+            accesses_psram: data_in_psram,
+        }),
+        data_len,
+    )
+}
+
+#[cfg(psram_dma)]
+fn build_descriptor_list_for_psram(
+    descriptors: &mut DescriptorSet<'_>,
+    copy_buffers: &mut [Option<ManualWritebackBuffer>; 2],
+    data: NonNull<[u8]>,
+) -> usize {
+    let data_len = data.len();
+    let data_addr = data.addr().get();
+
+    let min_alignment = ExternalBurstConfig::DEFAULT.min_psram_alignment(TransferDirection::In);
+    let chunk_size = 4096 - min_alignment;
+
+    let mut desciptor_iter = DescriptorChainingIter::new(descriptors.descriptors);
```

Comment:
Was gonna suggest simplifying `DescriptorChainingIter` to this.

```Rust
let mut desciptor_iter = descriptors.descriptors.iter_mut();

// Keep the desciptor_iter.next() stuff the same.

let unused = desciptor_iter.count(); // this is O(1)
let used =  descriptors.descriptors.len() - unused;

let next = null_mut();
for desc in descriptors.descriptors[..unused].iter_mut().rev() {
    desc.next = next;
    next = desc;
}
```

But the `return consumed;` complicates things.

Reply:
I suppose it could be done in `Drop` actually.

Reply:
Cryptic again, sorry 😅